pkg/virt-launcher/virtwrap/converter/converter.go

```diff
@@ -255,16 +261,65 @@ func Convert_v1_Disk_To_api_Disk(c *ConverterContext, diskDevice *v1.Disk, disk
 	return nil
 }
 
-func checkDirectIOFlag(path string) bool {
-	// check if fs where disk.img file is located or block device
-	// support direct i/o
+type DirectIOChecker interface {
+	CheckBlockDevice(path string) (bool, error)
+	CheckFile(path string) (bool, error)
+}
+
+type directIOChecker struct{}
+
+func NewDirectIOChecker() DirectIOChecker {
+	return &directIOChecker{}
+}
+
+func (c *directIOChecker) CheckBlockDevice(path string) (bool, error) {
+	return c.check(path, syscall.O_RDONLY)
+}
+
+func (c *directIOChecker) CheckFile(path string) (bool, error) {
+	flags := syscall.O_RDONLY
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		// try to create the file and perform the check
+		flags = flags | syscall.O_CREAT
+		defer os.Remove(path)
+	}
+	return c.check(path, flags)
+}
+
+// based on https://gitlab.com/qemu-project/qemu/-/blob/master/util/osdep.c#L344
+func (c *directIOChecker) check(path string, flags int) (bool, error) {
 	// #nosec No risk for path injection. No information can be exposed to attacker
-	f, err := os.OpenFile(path, syscall.O_RDONLY|syscall.O_DIRECT, 0)
-	if err != nil && !os.IsNotExist(err) {
-		return false
+	f, err := os.OpenFile(path, flags|syscall.O_DIRECT, 0600)
+	if err != nil {
+		// EINVAL is returned if the filesystem does not support the O_DIRECT flag
+		if err, ok := err.(*os.PathError); ok && err.Err == syscall.EINVAL {
+			// #nosec No risk for path injection. No information can be exposed to attacker
+			f, err := os.OpenFile(path, flags & ^syscall.O_DIRECT, 0600)
+			if err == nil {
+				defer util.CloseIOAndCheckErr(f, nil)
+				return false, nil
+			}
+		}
+		return false, err
 	}
 	defer util.CloseIOAndCheckErr(f, nil)
-	return true
+	return true, nil
+}
+
+// check if fs or block device support direct i/o
+func checkDirectIOFlag(path string, isBlockDev bool, checker DirectIOChecker) bool {
+	ok, err := func() (bool, error) {
+		if isBlockDev {
+			return checker.CheckBlockDevice(path)
+		} else {
+			return checker.CheckFile(path)
+		}
+	}()
+	if err != nil {
+		log.Log.Reason(err).Errorf("Direct IO check failed for %s", path)
+		return false
+	}
```

Comment:
why would we want to swallow the error here? If we fail to find an answer to the question "does this path support O_DIRECT", we probably have a pretty awkward situation (EIO? EPERM? I don't know). I don't understand why we should assume we have no direct IO and continue on. I think that it would be much safer if `SetDriverCacheMode` bubbles this error up (too bad GoLang does not have exceptions)

Reply:
This is done in order to preserve the existing behaviour. In general a failure in detection of direct IO support is not always a critical issue. Only when the VMI spec explicitly sets the cache mode to `none` and the detection fails (or direct IO is not supported) that is an error that needs to be propogated to upper layers. In other cases the `writethrough` cache policy can be used which does not require direct IO support. All that is already handled in `SetDriverCacheMode`.

Reply:
In my very humble non-definitive opinion, this complex behaviour is not worth preserving. If there's a case where we do not care about whether direct-io is needed or not, we should not check for it. If we do check for it, we should do something with the value. And if for some reason the caller of `checkDirectIOFlag` does want to ignore an obscure error, it can. I just find it wrong for a function to return a value that it is unsure about. A function should not assume anything about what its caller does.

Reply:
There is a case when we do care if direrct io is supported but if its not or we cannot check it because of an error we can still fallback to another option. Its when the VMI spec does not explicitly define the cache policy and we try to detect it. In that case a detection failure is not critical and we can go on with safe `writethrough` policy instead of completely refusing to start a VM. Printing a warning in that case might be usefull though.